File: test/core/data/repositories/supabase_task_repository_test.dart
import 'package:daypilot/core/cache/session_cache.dart';
import 'package:daypilot/core/data/models/app_task.dart';
import 'package:daypilot/core/data/models/task_category.dart';
import 'package:daypilot/core/data/models/task_difficulty.dart';
import 'package:daypilot/core/data/repositories/supabase_task_repository.dart';
import 'package:daypilot/core/utils/iso_date.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_test/flutter_test.dart';

import '../../../support/fake_supabase_http.dart';

void main() {
  late FakeSupabaseHttp http;
  late ProviderContainer container;
  late SupabaseTaskRepository repo;

  setUp(() {
    http = FakeSupabaseHttp();
    container = ProviderContainer();
    final repoProvider = Provider((ref) => SupabaseTaskRepository(http.client, ref));
    repo = container.read(repoProvider);
  });

  tearDown(() {
    container.dispose();
    return http.dispose();
  });

  group('getTasks', () {
    test('returns the cached list without making a request', () async {
      final cached = [
        AppTask(
          id: 't1',
          occurrenceId: 'o1',
          title: 'Cached',
          difficulty: TaskDifficulty.easy,
          category: TaskCategory.otro,
          date: DateTime(2026, 7, 25),
          durationMinutes: 10,
        ),
      ];
      container.read(tasksCacheProvider.notifier).state = cached;

      final tasks = await repo.getTasks();

      expect(tasks, cached);
      expect(http.requests, isEmpty);
    });

    test('with no signed-in user and no cache returns empty', () async {
      final tasks = await repo.getTasks();

      expect(tasks, isEmpty);
      expect(http.requests, isEmpty);
    });

    test('fetches, maps and caches tasks for the signed-in user', () async {
      await http.signIn('user-1');
      http.enqueue(jsonRows([
        {
          'task_id': 't1',
          'occurrence_id': 'o1',
          'title': 'Read',
          'description': null,
          'difficulty': 'MEDIUM',
          'category': 'Estudio',
          'date': '2026-07-25',
          'estimated_minutes': 30,
          'reminder_enabled': true,
          'is_recurring': false,
          'is_completed': false,
        },
      ]));

      final tasks = await repo.getTasks();

      expect(tasks, hasLength(1));
      expect(tasks.single.title, 'Read');
      expect(tasks.single.difficulty, TaskDifficulty.medium);
      expect(tasks.single.category, TaskCategory.estudio);
      expect(http.requests[0].url.queryParameters['user_id'], 'eq.user-1');

      // A second call should now be served from the cache the first call populated.
      final cachedTasks = await repo.getTasks();
      expect(cachedTasks, hasLength(1));
      expect(http.requests, hasLength(1));
    });
  });

  test('addTask with no signed-in user makes no request', () async {
    await repo.addTask(NewTaskData(
      date: DateTime(2026, 7, 25),
      title: 'Read',
      category: TaskCategory.estudio,
      difficulty: TaskDifficulty.easy,
      durationMinutes: 20,
    ));

    expect(http.requests, isEmpty);
  });

  test('addTask inserts the task and its single occurrence when not recurring', () async {
    await http.signIn('user-1');
    http.enqueue(noContent());
    http.enqueue(noContent());

    await repo.addTask(NewTaskData(
      date: DateTime(2026, 7, 25),
      title: 'Read',
      category: TaskCategory.estudio,
      difficulty: TaskDifficulty.hard,
      durationMinutes: 20,
    ));

    expect(http.requests, hasLength(2));
    expect(http.requests[0].url.path, '/rest/v1/tasks');
    expect(http.requests[0].body, contains('"title":"Read"'));
    expect(http.requests[0].body, contains('"category":"Estudio"'));
    expect(http.requests[0].body, contains('"difficulty":"HARD"'));
    expect(http.requests[1].url.path, '/rest/v1/task_days');
    expect(http.requests[1].body, contains('"date":"2026-07-25"'));
  });

  test('addTask expands a recurring task into occurrences up to 90 days out', () async {
    await http.signIn('user-1');
    for (var i = 0; i < 4; i++) {
      http.enqueue(noContent());
    }

    await repo.addTask(NewTaskData(
      date: DateTime(2026, 1, 1),
      title: 'Standup',
      category: TaskCategory.trabajo,
      difficulty: TaskDifficulty.easy,
      durationMinutes: 15,
      recurring: true,
      recurrenceDays: 45,
    ));

    // 1 tasks insert + 1 initial task_days insert + 2 recurring occurrences (day 45, day 90).
    expect(http.requests, hasLength(4));
    expect(http.requests[1].body, contains('"date":"${isoDate(DateTime(2026, 1, 1))}"'));
    expect(http.requests[2].body, contains('"date":"${isoDate(DateTime(2026, 1, 1).add(const Duration(days: 45)))}"'));
    expect(http.requests[3].body, contains('"date":"${isoDate(DateTime(2026, 1, 1).add(const Duration(days: 90)))}"'));
  });

  test('updateTask patches the editable fields scoped to the owner', () async {
    await http.signIn('user-1');
    http.enqueue(noContent());

    await repo.updateTask(
      id: 't1',
      title: 'New title',
      description: '',
      category: TaskCategory.salud,
      difficulty: TaskDifficulty.medium,
      durationMinutes: 25,
    );

    expect(http.requests[0].method, 'PATCH');
    expect(http.requests[0].url.path, '/rest/v1/tasks');
    expect(http.requests[0].url.queryParameters['id'], 'eq.t1');
    expect(http.requests[0].url.queryParameters['user_id'], 'eq.user-1');
    expect(http.requests[0].body, contains('"description":null'));
  });

  group('toggleTask', () {
    test('marking as not-done only updates task_days', () async {
      await http.signIn('user-1');
      http.enqueue(noContent());

      await repo.toggleTask(occurrenceId: 'o1', isDone: false);

      expect(http.requests, hasLength(1));
      expect(http.requests[0].body, contains('"is_completed":false'));
      expect(http.requests[0].body, contains('"completed_at":null'));
    });

    test('marking as done for the first time earns points', () async {
      await http.signIn('user-1');
      http.enqueue(noContent());
      http.enqueue(jsonRows([
        {'is_earned': false},
      ]));
      http.enqueue(noContent());
      http.enqueue(noContent());

      await repo.toggleTask(occurrenceId: 'o1', isDone: true);

      expect(http.requests, hasLength(4));
      expect(http.requests[2].url.path, '/rest/v1/points_log');
      expect(http.requests[2].body, contains('"points":20'));
      expect(http.requests[3].body, contains('"is_earned":true'));
    });

    test('marking as done when already earned does not award points again', () async {
      await http.signIn('user-1');
      http.enqueue(noContent());
      http.enqueue(jsonRows([
        {'is_earned': true},
      ]));

      await repo.toggleTask(occurrenceId: 'o1', isDone: true);

      expect(http.requests, hasLength(2));
    });
  });

  test('deleteTask deletes the row scoped to the owner', () async {
    await http.signIn('user-1');
    http.enqueue(noContent());

    await repo.deleteTask('t1');

    expect(http.requests[0].method, 'DELETE');
    expect(http.requests[0].url.path, '/rest/v1/tasks');
    expect(http.requests[0].url.queryParameters['id'], 'eq.t1');
    expect(http.requests[0].url.queryParameters['user_id'], 'eq.user-1');
  });
}

File: app/src/test/java/com/example/daypilot_test_desing/data/supabase/SupabaseTaskRepositoryTest.kt
package com.example.daypilot_test_desing.data.supabase

import com.example.daypilot_test_desing.core.cache.SessionCache
import com.example.daypilot_test_desing.core.data.model.NewTaskData
import com.example.daypilot_test_desing.core.data.model.TaskCategory
import com.example.daypilot_test_desing.core.data.model.TaskDifficulty
import com.example.daypilot_test_desing.support.SharedFakeSupabaseClient
import com.example.daypilot_test_desing.support.fakeLogin
import io.ktor.client.engine.mock.MockRequestHandleScope
import io.ktor.client.engine.mock.respond
import io.ktor.client.engine.mock.respondError
import io.ktor.client.engine.mock.toByteArray
import io.ktor.client.request.HttpRequestData
import io.ktor.client.request.HttpResponseData
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpStatusCode
import io.ktor.http.headersOf
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Before
import org.junit.Test

private val jsonHeaders = headersOf(HttpHeaders.ContentType, "application/json")

class SupabaseTaskRepositoryTest {

    companion object {
        private val shared = SharedFakeSupabaseClient()
    }

    private fun fakeSupabaseClient(
        handler: suspend MockRequestHandleScope.(HttpRequestData) -> HttpResponseData
    ) = shared.use(handler)

    @Before
    fun setUp() = runBlocking {
        shared.resetAuth()
        SessionCache.clear()
    }

    private fun taskRow(category: String, difficulty: String) = """{
        "occurrence_id":"o1","task_id":"t1","user_id":"u1","title":"T",
        "category":"$category","difficulty":"$difficulty",
        "is_completed":false,"estimated_minutes":30,"date":"2026-07-15"
    }"""

    @Test
    fun `getTasks returns the cached list without hitting the network`() = runTest {
        val cached = listOf(com.example.daypilot_test_desing.core.data.model.CalendarTaskData(
            id = "t1", occurrenceId = "o1", day = 1, month = 1, year = 2026,
            title = "Cached", category = TaskCategory.WORK, difficulty = TaskDifficulty.EASY,
            duration = 10, isDone = false
        ))
        SessionCache.tasks.value = cached
        val repo = SupabaseTaskRepository(fakeSupabaseClient { error("no HTTP expected") })

        assertEquals(cached, repo.getTasks())
    }

    @Test
    fun `getTasks returns empty when nobody is logged in`() = runTest {
        val repo = SupabaseTaskRepository(fakeSupabaseClient { error("no HTTP expected") })
        assertEquals(emptyList<Any>(), repo.getTasks())
    }

    @Test
    fun `getTasks maps every known category and difficulty label, and caches the result`() = runTest {
        val rows = listOf(
            "Estudio" to "MEDIUM", "Trabajo" to "HARD", "Deporte" to "EASY",
            "Salud" to "medium", "Bienestar" to "HARD", "General" to "EASY",
            "Hogar" to "MEDIUM", "Compra" to "HARD", "Finanzas" to "EASY",
            "SomethingElse" to "unknown"
        ).joinToString(",") { (cat, diff) -> taskRow(cat, diff) }
        val client = fakeSupabaseClient { respond("[$rows]", headers = jsonHeaders) }
        client.fakeLogin("u1")
        val repo = SupabaseTaskRepository(client)

        val result = repo.getTasks()

        assertEquals(10, result.size)
        assertEquals(
            listOf(
                TaskCategory.STUDY, TaskCategory.WORK, TaskCategory.SPORT,
                TaskCategory.HEALTH, TaskCategory.HEALTH, TaskCategory.OTHER,
                TaskCategory.HOME, TaskCategory.HOME, TaskCategory.WORK,
                TaskCategory.OTHER
            ),
            result.map { it.category }
        )
        assertEquals(
            listOf(
                TaskDifficulty.MEDIUM, TaskDifficulty.HARD, TaskDifficulty.EASY,
                TaskDifficulty.MEDIUM, TaskDifficulty.HARD, TaskDifficulty.EASY,
                TaskDifficulty.MEDIUM, TaskDifficulty.HARD, TaskDifficulty.EASY,
                TaskDifficulty.EASY
            ),
            result.map { it.difficulty }
        )
        assertEquals(2026, result.first().year)
        assertEquals(7, result.first().month)
        assertEquals(15, result.first().day)
        assertEquals(result, SessionCache.tasks.value)
    }

    @Test
    fun `getTasks returns empty when the request fails`() = runTest {
        val client = fakeSupabaseClient { respondError(HttpStatusCode.InternalServerError) }
        client.fakeLogin("u1")
        val repo = SupabaseTaskRepository(client)

        assertEquals(emptyList<Any>(), repo.getTasks())
    }

    private fun newTask(isRecurring: Boolean = false, recurrenceDays: Int = 1) = NewTaskData(
        day = 15, month = 7, year = 2026,
        title = "New task", category = TaskCategory.WORK, difficulty = TaskDifficulty.EASY,
        duration = 30, isRecurring = isRecurring, recurrenceDays = recurrenceDays
    )

    @Test
    fun `addTask is a no-op when nobody is logged in`() = runTest {
        val repo = SupabaseTaskRepository(fakeSupabaseClient { error("no HTTP expected") })
        repo.addTask(newTask())
    }

    @Test
    fun `addTask inserts the task then a single task_days row for a non-recurring task`() = runTest {
        var taskDaysInsertCount = 0
        val client = fakeSupabaseClient { request ->
            when {
                request.url.encodedPath.endsWith("/tasks") ->
                    respond("""[{"id":"real-id"}]""", headers = jsonHeaders)
                request.url.encodedPath.endsWith("/task_days") -> {
                    taskDaysInsertCount++
                    respond("[]", headers = jsonHeaders)
                }
                else -> respond("[]", headers = jsonHeaders)
            }
        }
        client.fakeLogin("u1")
        val repo = SupabaseTaskRepository(client)

        repo.addTask(newTask())

        assertEquals(1, taskDaysInsertCount)
        assertEquals(null, SessionCache.tasks.value)
    }

    @Test
    fun `addTask invalidates a previously cached task list`() = runTest {
        SessionCache.tasks.value = emptyList()
        val client = fakeSupabaseClient { request ->
            when {
                request.url.encodedPath.endsWith("/tasks") -> respond("""[{"id":"real-id"}]""", headers = jsonHeaders)
                else -> respond("[]", headers = jsonHeaders)
            }
        }
        client.fakeLogin("u1")
        val repo = SupabaseTaskRepository(client)

        repo.addTask(newTask())

        assertEquals(null, SessionCache.tasks.value)
    }

    @Test
    fun `addTask inserts recurring occurrences within the 90-day window`() = runTest {
        var bulkInsertSize = 0
        val client = fakeSupabaseClient { request ->
            when {
                request.url.encodedPath.endsWith("/tasks") ->
                    respond("""[{"id":"real-id"}]""", headers = jsonHeaders)
                request.url.encodedPath.endsWith("/task_days") -> {
                    val body = String(request.body.toByteArray())
                    // The bulk recurrence insert sends a JSON array; the first single insert doesn't.
                    if (body.trimStart().startsWith("[")) {
                        bulkInsertSize = body.split("task_id").size - 1
                    }
                    respond("[]", headers = jsonHeaders)
                }
                else -> respond("[]", headers = jsonHeaders)
            }
        }
        client.fakeLogin("u1")
        val repo = SupabaseTaskRepository(client)

        repo.addTask(newTask(isRecurring = true, recurrenceDays = 30))

        // Starting the day after the task's date, every 30 days, up to +90 days: 3 occurrences.
        assertEquals(3, bulkInsertSize)
    }

    @Test
    fun `addTask propagates the failure when the task insert fails`() = runTest {
        val client = fakeSupabaseClient { respondError(HttpStatusCode.InternalServerError) }
        client.fakeLogin("u1")
        val repo = SupabaseTaskRepository(client)

        try {
            repo.addTask(newTask())
            fail("expected an exception to propagate")
        } catch (_: Exception) {
            // expected
        }
    }

    @Test
    fun `updateTask is a no-op when nobody is logged in`() = runTest {
        val repo = SupabaseTaskRepository(fakeSupabaseClient { error("no HTTP expected") })
        repo.updateTask("t1", "New title", TaskCategory.WORK, TaskDifficulty.HARD, 45, "desc")
    }

    @Test
    fun `updateTask sends the new fields and invalidates the cache`() = runTest {
        SessionCache.tasks.value = emptyList()
        var captured: String? = null
        val client = fakeSupabaseClient { request ->
            captured = String(request.body.toByteArray())
            respond("[]", headers = jsonHeaders)
        }
        client.fakeLogin("u1")
        val repo = SupabaseTaskRepository(client)

        repo.updateTask("t1", "New title", TaskCategory.STUDY, TaskDifficulty.HARD, 45, "desc")

        assertTrue(captured!!.contains("\"title\":\"New title\""))
        assertTrue(captured!!.contains("\"difficulty\":\"HARD\""))
        assertEquals(null, SessionCache.tasks.value)
    }

    @Test
    fun `updateTask propagates the failure`() = runTest {
        val client = fakeSupabaseClient { respondError(HttpStatusCode.InternalServerError) }
        client.fakeLogin("u1")
        val repo = SupabaseTaskRepository(client)

        try {
            repo.updateTask("t1", "T", TaskCategory.WORK, TaskDifficulty.EASY, 10, "")
            fail("expected an exception to propagate")
        } catch (_: Exception) {
            // expected
        }
    }

    @Test
    fun `toggleTask is a no-op when nobody is logged in`() = runTest {
        val repo = SupabaseTaskRepository(fakeSupabaseClient { error("no HTTP expected") })
        repo.toggleTask("o1", true)
    }

    @Test
    fun `toggleTask marking done sets is_earned and a completed_at timestamp`() = runTest {
        var captured: String? = null
        val client = fakeSupabaseClient { request ->
            captured = String(request.body.toByteArray())
            respond("[]", headers = jsonHeaders)
        }
        client.fakeLogin("u1")
        val repo = SupabaseTaskRepository(client)

        repo.toggleTask("o1", true)

        assertTrue(captured!!.contains("\"is_completed\":true"))
        assertTrue(captured!!.contains("\"is_earned\":true"))
        assertTrue(captured!!.contains("\"completed_at\":\""))
    }

    @Test
    fun `toggleTask marking pending clears completed_at and omits is_earned`() = runTest {
        var captured: String? = null
        val client = fakeSupabaseClient { request ->
            captured = String(request.body.toByteArray())
            respond("[]", headers = jsonHeaders)
        }
        client.fakeLogin("u1")
        val repo = SupabaseTaskRepository(client)

        repo.toggleTask("o1", false)

        assertTrue(captured!!.contains("\"is_completed\":false"))
        assertTrue(captured!!.contains("\"completed_at\":null"))
        assertTrue(!captured!!.contains("is_earned"))
    }

    @Test
    fun `toggleTask propagates the failure`() = runTest {
        val client = fakeSupabaseClient { respondError(HttpStatusCode.InternalServerError) }
        client.fakeLogin("u1")
        val repo = SupabaseTaskRepository(client)

        try {
            repo.toggleTask("o1", true)
            fail("expected an exception to propagate")
        } catch (_: Exception) {
            // expected
        }
    }

    @Test
    fun `deleteTask is a no-op when nobody is logged in`() = runTest {
        val repo = SupabaseTaskRepository(fakeSupabaseClient { error("no HTTP expected") })
        repo.deleteTask("t1")
    }

    @Test
    fun `deleteTask removes the row and invalidates the cache`() = runTest {
        SessionCache.tasks.value = emptyList()
        val client = fakeSupabaseClient { request ->
            assertTrue(request.url.encodedPath.endsWith("/tasks"))
            respond("[]", headers = jsonHeaders)
        }
        client.fakeLogin("u1")
        val repo = SupabaseTaskRepository(client)

        repo.deleteTask("t1")

        assertEquals(null, SessionCache.tasks.value)
    }

    @Test
    fun `deleteTask propagates the failure`() = runTest {
        val client = fakeSupabaseClient { respondError(HttpStatusCode.InternalServerError) }
        client.fakeLogin("u1")
        val repo = SupabaseTaskRepository(client)

        try {
            repo.deleteTask("t1")
            fail("expected an exception to propagate")
        } catch (_: Exception) {
            // expected
        }
    }
}
